### PR TITLE
Added Time Distributed Softmax to allow softmax activation over 3D inputs

### PR DESCRIPTION
**Pull request by patyork**
Allows softmax activation over tensor3 values. Given output A from this activation, `sum(A[i][j][:])==1.0`

Impetus: Output from RNNs is generally considered of size `(nb_samples, time_steps, values)`. This activation allows for softmax over this tensor3, where the softmax is applied over a vector of size `(values, )` which corresponds to the softmax of one sample at one time step.

Example:
```python
model = Sequential()
model.add(SimpleRNN(2, 3, return_sequences=True))
model.add(Activation('time_distributed_softmax'))

sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
model.compile(loss='mse', optimizer='sgd')
A = model._predict(np.asarray([[[1,1],[2,2]], [[3,3], [4,4]]]))

print A
print np.sum(A[0][0][:])
```

Yields
```python
[[[ 0.32846826  0.33231813  0.33921361]
  [ 0.32311559  0.33156598  0.34531844]]

 [[ 0.31882143  0.33012739  0.35105121]
  [ 0.31376582  0.32922304  0.35701114]]]
1.0
```

### CODE REVIEW

**Comment by patyork** [comment]
Also, additional name ideas:
* Time Sensitive Softmax
* Dimension Sensitive Softmax
* Shape Sensitive Softmax
* Softmax 3D

From Torch, they have just a single SoftMax function, which works in the same way for n-dimensional tensors, such that the sum of the last dimension, for a single subtensor, sums to 1. https://github.com/torch/nn/blob/master/doc/transfer.md

I think we should still bounce around ideas for making SoftMax perform as Torch does (Which would be the dimensionality check in the SoftMax activation, a reshape if required to 2D, application of softmax, and a reshape back if required).

**Comment by nagadomi** [comment]
Do we need a fully customizable reshape?
```python
import numpy as np
import theano
from keras.models import Sequential
from keras.layers.core import Activation, Layer
from keras.optimizers import SGD
from keras.utils import np_utils, generic_utils
from keras.layers.recurrent import SimpleRNN
from keras.utils.generic_utils import make_tuple

class MyReshape(Layer):
    def __init__(self, dims, batchwise = True):
        self.dims = dims
        self.params = []
        self.batchwise = batchwise
        
    def output(self, train):
        X = self.get_input(train)
        if self.batchwise:
            nshape = make_tuple(X.shape[0], *self.dims)
        else:
            nshape = make_tuple(*self.dims)
        return theano.tensor.reshape(X, nshape)

model = Sequential()
model.add(SimpleRNN(2, 3, return_sequences=True))
model.add(MyReshape((2 * 2, 3), batchwise=False))
model.add(Activation('softmax'))
# model.add(MyReshape((2, 2, 3), batchwise=False)) # maybe this line breaks the metric of loss function.

sgd = SGD(lr=0.1, decay=1e-6, momentum=0.9, nesterov=True)
model.compile(loss='mse', optimizer='sgd')

x = np.asarray([[[1,1],[2,2]], [[3,3], [4,4]]])
A = model._predict(x).reshape(2, 2, 3) # reshape

print A
print np.sum(A[0][0][:])
```
```
[[[ 0.33217808  0.33681527  0.33100665]
  [ 0.33198774  0.33990619  0.32810607]]

 [[ 0.32982884  0.34381771  0.32635345]
  [ 0.32959331  0.34696015  0.32344653]]]
1.0
```
I don't know much about RNN.

**Comment by patyork** [comment]
The above is a little off, it shouldn't compile (3 output -> 2 * 2 input).

```python
model = Sequential()
model.add(SimpleRNN(2, 3, return_sequences=True))
model.add(MyReshape((3 * 4, 5), batchwise=False))
model.add(Activation('softmax'))
```
This makes the assumption that the batch size will be a constant 4. This may break if, for example we have a training set of 14 samples, the last batch would only be of size 2. Thus the reshape would break.

The Reshape layer seems a little buggy to me - I can't really think of a model it which it would be useful, especially with the first dimension fixed, and no way to revert back and forth from, say, 2D -> 3D -> 2D (which has to happen for RNNs to be effective, or even possible). To go back and forth like that, you need to know the batch size and the input size, at a minimum, to be able to calculate the reshape intelligently and correctly.

With RNNs, you need to utilize tensor3 objects to get the job done. For example, with a batch input:
```python
# Forward recurrent network, where outputs of recurrent layer at T1 are also inputs to the recurrent layer at T2
output = []
for sample in batch:
    previous = np.zeros((.. , ..))
    for time_step in sample:
        previous = concatenate( dot(previous, W1), dot(time_step, W2)) + b
    output.append(previous)
return output
```
The above double loop doesn't make sense - parallelism is key. So we can work some magic:
```python
reshaped_input = input.reshape(timesteps, batch_size, values)
previous = np.zeros((.. , ..))
for time_step in reshaped_input:
    previous = concatenate( dot(previous, W1), dot(time_step, W2)) + b
return previous.reshape(batch_size, previous.shape[0]/batch_size, output_size
or
return previous.reshape(previous.shape[0]/batch_size, batch_size, output_size
```
...where the first return gives a tensor3 out, or the second gives a restacked matrix, for faster parallel processing.

The problem with the above is that it assumes that a tensor3 object is the type of the input.

**Comment by nagadomi** [comment]
oh, I see.
